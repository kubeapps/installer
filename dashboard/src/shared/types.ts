import { IAuthState } from "../reducers/auth";
import { IServiceCatalogState } from "../reducers/catalog";
import { IConfigState } from "../reducers/config";
import { IFunctionState } from "../reducers/functions";
import { INamespaceState } from "../reducers/namespace";
import { IAppRepositoryState } from "../reducers/repos";
import { hapi } from "./hapi/release";

// Allow defining multiple error classes
// tslint:disable:max-classes-per-file
export class ForbiddenError extends Error {
  constructor(message?: string) {
    super(message);
    Object.setPrototypeOf(this, new.target.prototype);
  }
}

export class NotFoundError extends Error {
  constructor(message?: string) {
    super(message);
    Object.setPrototypeOf(this, new.target.prototype);
  }
}

export class MissingChart extends Error {
  constructor(message?: string) {
    super(message);
    Object.setPrototypeOf(this, new.target.prototype);
  }
}

export class AppConflict extends Error {
  constructor(message?: string) {
    super(message);
    Object.setPrototypeOf(this, new.target.prototype);
  }
}

export interface IRepo {
  name: string;
  url: string;
}

export interface IChartVersion {
  id: string;
  attributes: IChartVersionAttributes;
  relationships: {
    chart: {
      data: IChartAttributes;
    };
  };
}

export interface IChartVersionAttributes {
  version: string;
  app_version: string;
  created: string;
}

export interface IChart {
  id: string;
  attributes: IChartAttributes;
  relationships: {
    latestChartVersion: {
      data: IChartVersionAttributes;
    };
  };
}

export interface IChartAttributes {
  name: string;
  description: string;
  home?: string;
  icon?: string;
  keywords: string[];
  maintainers: Array<{
    name: string;
    email?: string;
  }>;
  repo: IRepo;
  sources: string[];
}

export interface IChartState {
  isFetching: boolean;
  selected: {
    error?: Error;
    version?: IChartVersion;
    versions: IChartVersion[];
    readme?: string;
    readmeError?: string;
    values?: string;
  };
  items: IChart[];
}

export interface IDeployment {
  metadata: {
    name: string;
    namespace: string;
  };
}

export interface IServiceSpec {
  ports: IPort[];
  clusterIP: string;
  type: string;
}

export interface IServiceStatus {
  loadBalancer: {
    ingress?: Array<{ ip: string }>;
  };
}

export interface IPort {
  name: string;
  port: number;
  protocol: string;
  targetPort: string;
  nodePort: string;
}

export interface IResource {
  apiVersion: string;
  kind: string;
  type: string;
  spec: any;
  status: any;
  metadata: {
    name: string;
    namespace: string;
    annotations: string;
    creationTimestamp: string;
    selfLink: string;
    resourceVersion: string;
    deletionTimestamp?: string;
    uid: string;
  };
}

export interface IOwnerReference {
  apiVersion: string;
  blockOwnerDeletion: boolean;
  kind: string;
  name: string;
  uid: string;
}

export interface ISecret extends IResource {
  data: { [s: string]: string };
}

export interface IDeploymentStatus {
  replicas: number;
  updatedReplicas: number;
  availableReplicas: number;
}

export interface IFunction extends IResource {
  spec: {
    deps: string;
    function: string;
    handler: string;
    runtime: string;
    checksum: string;
  };
}

export interface IAppState {
  isFetching: boolean;
  error?: Error;
  deleteError?: Error;
  // currently items are always Helm releases
  items: hapi.release.Release[];
  listOverview?: IAppOverview[];
  selected?: hapi.release.Release;
}

export interface IStoreState {
  catalog: IServiceCatalogState;
  apps: IAppState;
  auth: IAuthState;
  charts: IChartState;
  config: IConfigState;
  repos: IAppRepositoryState;
  deployment: IDeployment;
  functions: IFunctionState;
  namespace: INamespaceState;
}

interface IK8sResource {
  apiVersion: string;
  kind: string;
}

/** @see https://github.com/kubernetes/community/blob/master/contributors/devel/api-conventions.md#objects */
export interface IK8sObject<M, SP, ST> extends IK8sResource {
  metadata: {
    annotations?: { [key: string]: string };
    creationTimestamp?: string;
    deletionTimestamp?: string | null;
    generation?: number;
    labels?: { [key: string]: string };
    name: string;
    namespace: string;
    resourceVersion?: string;
    uid: string;
    selfLink?: string; // Not in docs, but seems to exist everywhere
  } & M;
  spec?: SP;
  status?: ST;
}

/** @see https://github.com/kubernetes/community/blob/master/contributors/devel/api-conventions.md#lists-and-simple-kinds */
export interface IK8sList<I, M> extends IK8sResource {
  items: I[];
  metadata?: {
    resourceVersion?: string;
    selfLink?: string; // Not in docs, but seems to exist everywhere
  } & M;
}

export interface IAppRepository
  extends IK8sObject<
      {
        clusterName: string;
        creationTimestamp: string;
        deletionGracePeriodSeconds: string | null;
        deletionTimestamp: string | null;
        resourceVersion: string;
        selfLink: string;
      },
      {
        type: string;
        url: string;
        auth: {
          header: {
            secretKeyRef: {
              name: string;
              key: string;
            };
          };
        };
      },
      undefined
    > {}

export interface IAppRepositoryList
  extends IK8sList<
      IAppRepository,
      {
        continue: string;
        resourceVersion: string;
        selfLink: string;
      }
    > {}

export interface IFunctionList extends IK8sList<IFunction, {}> {}

/** @see https://github.com/kubernetes/community/blob/master/contributors/devel/api-conventions.md#response-status-kind */
export interface IStatus extends IK8sResource {
  kind: "Status";
  status: "Success" | "Failure";
  message: string;
  reason:
    | "BadRequest"
    | "Unauthorized"
    | "Forbidden"
    | "NotFound"
    | "AlreadyExists"
    | "Conflict"
    | "Invalid"
    | "Timeout"
    | "ServerTimeout"
    | "MethodNotAllowed"
    | "InternalError";
  details?: {
    kind?: string;
    name?: string;
    causes?: IStatusCause[] | string;
  };
}

interface IStatusCause {
  field: string;
  message: string;
  reason: string;
}

export interface IRouterPathname {
  router: {
    location: {
      pathname: string;
    };
  };
}

export interface IKubelessConfigMap {
  metadata: {
    name: string;
  };
  data: {
    "runtime-images": string;
  };
}

export interface IRuntimeVersion {
  name: string;
  version: string;
  runtimeImage: string;
  initImage: string;
}

export interface IRuntime {
  ID: string;
  versions: IRuntimeVersion[];
  depName: string;
  fileNameSuffix: string;
}

export interface IRBACRole {
  apiGroup: string;
  namespace?: string;
  clusterWide?: boolean;
  resource: string;
  verbs: string[];
}

export interface IAppOverview {
  releaseName: string;
  namespace: string;
  version: string;
  icon?: string;
}
