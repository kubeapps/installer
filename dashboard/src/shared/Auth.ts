import Axios, { AxiosRequestConfig } from "axios";

const AuthTokenKey = "kubeapps_auth_token";

export class Auth {
  public static getAuthToken() {
    return localStorage.getItem(AuthTokenKey);
  }

  public static setAuthToken(token: string) {
    return localStorage.setItem(AuthTokenKey, token);
  }

  public static wsProtocols() {
    const token = this.getAuthToken();
    if (!token) {
      return [];
    }
    return ["base64url.bearer.authorization.k8s.io." + btoa(token).replace(/=*$/g, "")];
  }

  public static fetchOptions() {
    return { headers: { Authorization: `Bearer ${this.getAuthToken()}` } };
  }
}

// authenticatedAxiosInstance returns an axios instance with an interceptor
// configured to set the current auth token
function authenticatedAxiosInstance() {
  const a = Axios.create();
  a.interceptors.request.use((config: AxiosRequestConfig) => {
    const authToken = Auth.getAuthToken();
    if (authToken) {
      config.headers.Authorization = `Bearer ${authToken}`;
    }
    return config;
  });
  return a;
}

export const axios = authenticatedAxiosInstance();
