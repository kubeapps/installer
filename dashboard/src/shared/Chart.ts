import axios from "axios";

// import { IFunction, IFunctionList, IResource, IStatus } from "./types";

export default class Chart {
  public static async getReadme(id: string, version: string) {
    const url = `${Chart.APIEndpoint}/assets/${id}/versions/${version}/README.md`;
    const { data } = await axios.get<string>(url);
    return data;
  }

  public static async getValues(id: string, version: string) {
    const url = `${Chart.APIEndpoint}/assets/${id}/versions/${version}/values.yaml`;
    const { data } = await axios.get<string>(url);
    return data;
  }

  private static APIEndpoint: string = "/api/chartsvc/v1";
}
