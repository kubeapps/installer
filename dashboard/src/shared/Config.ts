import axios from "axios";

// IConfig is the configuration for Kubeapps
export interface IConfig {
  namespace: string;
}

export default class Config {
  public static async getConfig() {
    const url = Config.APIEndpoint;
    const { data } = await axios.get<IConfig>(url);
    return data;
  }

  private static APIEndpoint: string = "/config.json";
}
