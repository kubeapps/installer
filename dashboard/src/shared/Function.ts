import axios from "axios";

import { IFunction, IFunctionList } from "./types";

export default class Function {
  public static async list() {
    const { data } = await axios.get<IFunctionList>(`${Function.APIEndpoint}/functions`);
    return data;
  }

  public static async get(name: string, namespace: string) {
    const { data } = await axios.get(Function.getSelfLink(name, namespace));
    return data;
  }

  public static async update(name: string, namespace: string, newFn: IFunction) {
    const { data } = await axios.put(Function.getSelfLink(name, namespace), newFn);
    return data;
  }

  public static async delete(name: string, namespace: string) {
    const { data } = await axios.delete(Function.getSelfLink(name, namespace));
    return data;
  }

  //   public static async create(name: string, url: string) {
  //   }

  private static APIEndpoint: string = "/api/kube/apis/kubeless.io/v1beta1";
  private static getSelfLink(name: string, namespace: string): string {
    return `${Function.APIEndpoint}/namespaces/${namespace}/functions/${name}`;
  }
}
