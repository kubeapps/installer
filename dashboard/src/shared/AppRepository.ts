import { axios } from "./Auth";
import Config from "./Config";
import { IAppRepository, IAppRepositoryList } from "./types";

export class AppRepository {
  public static async list() {
    const { data } = await axios.get<IAppRepositoryList>(await AppRepository.getResourceLink());
    return data;
  }

  public static async get(name: string) {
    const { data } = await axios.get(await AppRepository.getResourceLink(name));
    return data;
  }

  public static async update(name: string, newApp: IAppRepository) {
    const { data } = await axios.put(await AppRepository.getResourceLink(name), newApp);
    return data;
  }

  public static async delete(name: string) {
    const { data } = await axios.delete(await AppRepository.getResourceLink(name));
    return data;
  }

  public static async create(name: string, url: string, auth: any) {
    const { data } = await axios.post<IAppRepository>(await AppRepository.getResourceLink(), {
      apiVersion: "kubeapps.com/v1alpha1",
      kind: "AppRepository",
      metadata: {
        name,
      },
      spec: { auth, type: "helm", url },
    });
    return data;
  }

  private static APIBase: string = "/api/kube";
  private static APIEndpoint: string = `${AppRepository.APIBase}/apis/kubeapps.com/v1alpha1`;
  private static async getResourceLink(name?: string) {
    const { namespace } = await Config.getConfig();
    return `${AppRepository.APIEndpoint}/namespaces/${namespace}/apprepositories${
      name ? `/${name}` : ""
    }`;
  }
}
