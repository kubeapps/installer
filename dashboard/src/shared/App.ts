import { AppRepository } from "./AppRepository";
import { axios } from "./Auth";
import { hapi } from "./hapi/release";
import { IAppOverview, IChartVersion } from "./types";

export class App {
  public static getResourceURL(namespace?: string, name?: string) {
    let url = "/api/tiller-deploy/v1";
    if (namespace) {
      url += `/namespaces/${namespace}`;
    }
    url += "/releases";
    if (name) {
      url += `/${name}`;
    }
    return url;
  }

  public static async create(
    releaseName: string,
    namespace: string,
    kubeappsNamespace: string,
    chartVersion: IChartVersion,
    values?: string,
  ) {
    const chartAttrs = chartVersion.relationships.chart.data;
    const repo = await AppRepository.get(chartAttrs.repo.name, kubeappsNamespace);
    const auth = repo.spec.auth;
    const endpoint = App.getResourceURL(namespace);
    const { data } = await axios.post(endpoint, {
      auth,
      chartName: chartAttrs.name,
      releaseName,
      repoUrl: chartAttrs.repo.url,
      values,
      version: chartVersion.attributes.version,
    });
    return data;
  }

  public static async upgrade(
    releaseName: string,
    namespace: string,
    kubeappsNamespace: string,
    chartVersion: IChartVersion,
    values?: string,
  ) {
    const chartAttrs = chartVersion.relationships.chart.data;
    const repo = await AppRepository.get(chartAttrs.repo.name, kubeappsNamespace);
    const auth = repo.spec.auth;
    const endpoint = App.getResourceURL(namespace, releaseName);
    const { data } = await axios.put(endpoint, {
      auth,
      chartName: chartAttrs.name,
      releaseName,
      repoUrl: chartAttrs.repo.url,
      values,
      version: chartVersion.attributes.version,
    });
    return data;
  }

  public static async delete(releaseName: string, namespace: string) {
    const { data } = await axios.delete(App.getResourceURL(namespace, releaseName));
    return data;
  }

  public static async listApps(namespace?: string) {
    const { data } = await axios.get<{ data: IAppOverview[] }>(App.getResourceURL(namespace));
    return data.data;
  }

  public static async getRelease(namespace: string, name: string) {
    const { data } = await axios.get<{ data: hapi.release.Release }>(
      this.getResourceURL(namespace, name),
    );
    return data.data;
  }
}
