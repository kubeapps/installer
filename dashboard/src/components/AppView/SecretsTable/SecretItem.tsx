import { isEmpty } from "lodash";
import * as React from "react";
import { AlertTriangle } from "react-feather";

import LoadingWrapper, { LoaderType } from "../../../components/LoadingWrapper";
import { IKubeItem, ISecret } from "../../../shared/types";
import "./SecretContent.css";
import SecretItemDatum from "./SecretItemDatum";

interface ISecretItemProps {
  name: string;
  secret?: IKubeItem<ISecret>;
  getSecret: () => void;
}

class SecretItem extends React.Component<ISecretItemProps> {
  public componentDidMount() {
    this.props.getSecret();
  }

  public render() {
    const { name, secret } = this.props;
    return (
      <tr className="flex">
        <td className="col-3">{name}</td>
        {this.renderSecretInfo(secret)}
      </tr>
    );
  }

  private renderSecretInfo(secret?: IKubeItem<ISecret>) {
    if (secret === undefined || secret.isFetching) {
      return (
        <td className="col-9">
          <LoadingWrapper type={LoaderType.Placeholder} />
        </td>
      );
    }
    if (secret.error) {
      return (
        <td className="col-9">
          <span className="flex">
            <AlertTriangle />
            <span className="flex margin-l-normal">Error: {secret.error.message}</span>
          </span>
        </td>
      );
    }
    if (secret.item) {
      const item = secret.item;
      return (
        <React.Fragment>
          <td className="col-2">{item.type}</td>
          {isEmpty(item.data) ? (
            <td className="col-7">
              <span>This Secret is empty</span>
            </td>
          ) : (
            <td className="col-7 padding-small">
              {Object.keys(item.data).map(k => (
                <SecretItemDatum key={`${item.metadata.name}/${k}`} name={k} value={item.data[k]} />
              ))}
            </td>
          )}
        </React.Fragment>
      );
    }
    return null;
  }
}

export default SecretItem;
