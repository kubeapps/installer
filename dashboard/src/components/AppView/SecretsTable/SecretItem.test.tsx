import { shallow } from "enzyme";
import context from "jest-plugin-context";
import * as React from "react";

import itBehavesLike from "../../../shared/specs";
import { ISecret } from "../../../shared/types";
import SecretItem from "./SecretItem";
import SecretItemDatum from "./SecretItemDatum";

describe("componentDidMount", () => {
  it("calls getSecret", () => {
    const mock = jest.fn();
    shallow(<SecretItem name="foo" getSecret={mock} />);
    expect(mock).toHaveBeenCalled();
  });
});

context("when fetching secrets", () => {
  [undefined, { isFetching: true }].forEach(secret => {
    itBehavesLike("aLoadingComponent", {
      component: SecretItem,
      props: {
        secret,
        getSecret: jest.fn(),
      },
    });
    it("displays the name of the Secret", () => {
      const wrapper = shallow(<SecretItem secret={secret} name="foo" getSecret={jest.fn()} />);
      expect(wrapper.text()).toContain("foo");
    });
  });
});

context("when there is an error fetching the Secret", () => {
  const secret = {
    error: new Error('secrets "foo" not found'),
    isFetching: false,
  };
  const wrapper = shallow(<SecretItem secret={secret} name="foo" getSecret={jest.fn()} />);

  it("diplays the Service name in the first column", () => {
    expect(
      wrapper
        .find("td")
        .first()
        .text(),
    ).toEqual("foo");
  });

  it("displays the error message in the second column", () => {
    expect(
      wrapper
        .find("td")
        .at(1)
        .text(),
    ).toContain('Error: secrets "foo" not found');
  });
});

context("when there is a valid Secret", () => {
  const secret = {
    metadata: {
      name: "foo",
    },
    type: "Opaque",
    data: {
      foo: "YmFy", // bar
      foo2: "YmFyMg==", // bar2
    } as { [s: string]: string },
  } as ISecret;
  const kubeItem = {
    isFetching: false,
    item: secret,
  };

  it("renders the Secret name and type", () => {
    const wrapper = shallow(<SecretItem secret={kubeItem} name="foo" getSecret={jest.fn()} />);
    expect(
      wrapper
        .find("td")
        .at(0)
        .text(),
    ).toContain("foo");
    expect(
      wrapper
        .find("td")
        .at(1)
        .text(),
    ).toContain("Opaque");
    expect(wrapper).toMatchSnapshot();
  });

  it("renders a SecretItemDatum component for each Secret key", () => {
    const wrapper = shallow(<SecretItem secret={kubeItem} name="foo" getSecret={jest.fn()} />);
    expect(
      wrapper
        .find("td")
        .at(2)
        .find(SecretItemDatum),
    ).toHaveLength(2);
    expect(
      wrapper
        .find(SecretItemDatum)
        .first()
        .props(),
    ).toMatchObject({ name: "foo", value: "YmFy" });
  });
});

context("when there is an empty Secret", () => {
  const secret = {
    metadata: {
      name: "foo",
    },
    type: "Opaque",
  } as ISecret;
  const kubeItem = {
    isFetching: false,
    item: secret,
  };

  it("displays a message", () => {
    const wrapper = shallow(<SecretItem secret={kubeItem} name="foo" getSecret={jest.fn()} />);
    expect(
      wrapper
        .find("td")
        .at(0)
        .text(),
    ).toContain("foo");
    expect(
      wrapper
        .find("td")
        .at(1)
        .text(),
    ).toContain("Opaque");
    expect(
      wrapper
        .find("td")
        .at(2)
        .text(),
    ).toContain("This Secret is empty");
    expect(wrapper).toMatchSnapshot();
  });
});
