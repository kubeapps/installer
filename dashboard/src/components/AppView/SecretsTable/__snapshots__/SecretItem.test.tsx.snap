// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`when fetching secrets loading spinner matches the snapshot 1`] = `
<tr
  className="flex"
>
  <td
    className="col-3"
  />
  <td
    className="col-9"
  >
    <LoadingWrapper
      loaded={false}
      type={1}
    />
  </td>
</tr>
`;

exports[`when fetching secrets loading spinner matches the snapshot 2`] = `
<tr
  className="flex"
>
  <td
    className="col-3"
  />
  <td
    className="col-9"
  >
    <LoadingWrapper
      loaded={false}
      type={1}
    />
  </td>
</tr>
`;

exports[`when there is a valid Secret renders the Secret name and type 1`] = `
<tr
  className="flex"
>
  <td
    className="col-3"
  >
    foo
  </td>
  <td
    className="col-2"
  >
    Opaque
  </td>
  <td
    className="col-7 padding-small"
  >
    <SecretItemDatum
      key="foo/foo"
      name="foo"
      value="YmFy"
    />
    <SecretItemDatum
      key="foo/foo2"
      name="foo2"
      value="YmFyMg=="
    />
  </td>
</tr>
`;

exports[`when there is an empty Secret displays a message 1`] = `
<tr
  className="flex"
>
  <td
    className="col-3"
  >
    foo
  </td>
  <td
    className="col-2"
  >
    Opaque
  </td>
  <td
    className="col-7"
  >
    <span>
      This Secret is empty
    </span>
  </td>
</tr>
`;
